src/packet.c

```diff
@@ -1,3 +1,153 @@
-int parse() { return 0; }
+#include "packet.h"
+#include "util.h"
 
-int serialize() { return 0; }
+struct SizeLUT {
+    uint32_t size;
+    void (*cb)(void);
+};
+
+static struct SizeLUT SERIAL_LUT[] = {
+    [BCMP_PARSER_8BIT] = { sizeof(uint8_t), NULL },
+    [BCMP_PARSER_16BIT] = { sizeof(uint16_t), swap_16bit },
+    [BCMP_PARSER_32BIT] = { sizeof(uint32_t), swap_32bit },
+    [BCMP_PARSER_64BIT] = { sizeof(uint64_t), swap_64bit },
+};
+
+static struct Parser {
+    LL ll;
+    uint32_t c;
+} PARSER;
+
+
+/*!
+ @brief Handler For Parser Traverse Callback Function
+
+ @param buf configuration of the message item
+ @param arg data to be processed by the message
+
+ @return BmOK on success
+ @return BmError on failure
+ */
+static BMErr parser_handler(void *buf, void *arg) {
+    BMErr err = BmEINVAL;
+    BCMPMessageParserCfg *cfg = NULL;
+    BCMPParserData data;
+
+    if (arg && buf) {
+        err = BmOK;
+        cfg = (BCMPMessageParserCfg *) buf;
+        data = *(BCMPParserData *) arg;
+        for (uint32_t i = 0; i < cfg->msg.count; i++) {
+            if (data.header->type == cfg->msg.types[i]) {
+               if (cfg->cb && cfg->cb(data) != BmOK) {
+                   //TODO log something clever here
+               }
+               else {
+                   break;
+               }
+           }
+        }
+    }
+
+    return err;
+}
+
+/*!
+ @brief Initialize Packet Parser
+
+ @return BmOK
+ */
+BMErr bcmp_parser_init(void) {
+    memset(PARSER, 0, sizeof(struct Parser));
+    return BmOK;
+}
+
+/*!
+ @brief Add Item To Be Parsed When Handling Incoming Messages
+
+ @details This will be a linked list item that is serviced in
+          the update function
+
+ @param item parser item to add to linked list
+
+ @return BmOK on success
+ @return BmError on failure
+ */
+BMErr bcmp_parser_add(BCMPParserItem *item) {
+    BMErr err = BmEINVAL;
+    LLItem *object = item->object;
+
+    if (item) {
+        object->data = &item->Cfg;
+        object->id = PARSER.c++;
+        err = ll_item_add(&PARSER.ll object);
+    }
+
+    return err;
+}
+
+/*!
+ @brief Update Function To Parse And Handle Incoming Message
+
+ @details This is to be ran on incoming messages, the function
+          serializes the header from the received message and then
+          reports the payload to the associated parser item. The parser
+          is then responsible for serializing the payload.
+
+ @param pbuf lwip pbuf incoming message
+ @param src ip address of source message
+ @param dst ip address of destination message
+
+ @return BmOK on success
+ @return BmError on failure
+ */
+BMErr bcmp_parse_update(struct pbuf *pbuf, ip_addr_t *src, ip_addr_t *dst) {
```

Comment:
With this, @towynlin I was still utilizing the lwip variables.
One way around this that I thought of, was instantiating a callback that allow us to format the data received over the line to avoid relying on lwip, but seeing how `stuct pbuf` is so tightly ingrained in the rest of the system, and other function calls associated with processing incoming messages, I decided to keep this as is.